src/snitch_section.cpp

```diff
@@ -15,12 +16,20 @@ section_entry_checker::~section_entry_checker() {
             // We are unwinding the stack because an exception has been thrown;
             // avoid touching the section state since we will want to report where
             // the exception was thrown.
+            state.reg.report_callback(
+                state.reg, event::section_ended{
+                               state.sections.current_section.back().id,
+                               state.sections.current_section.back().location, true});
             return;
         }
 #endif
 
         pop_location(state);
 
+        asserts          = state.asserts - asserts;
+        failures         = state.failures - failures;
+        allowed_failures = state.allowed_failures - allowed_failures;
```

Comment:
This gives two different meanings to `section_entry_checker::asserts`/`failures`/`allowed_failures` depending on context: on creation they hold the initial state, and on destruction they hold the difference. This will be a source of confusion, I fear. But in fact, we only need to store the initial state (and should probably rename the member variables as such, e.g. `initial_sate.asserts`/`initial_state.failures`/...); the difference computed here could be stored in local variables.

Reply:
> Alternatively, we could instead store the actual number of assertions (etc) that were recorded, and make the registry propagate the counts to all open sections in register_assertion(). This is a little bit more work for the CPU, but the code might be simpler.

This also might be necessary since I am noticing in my tests that Catch2 counts these things cumulatively for nested sections but with this strategy snitch does not.

given
```c++
SECTION("Section1") {
    CHECK(true);
    SECTION("Section1.1") {
        CHECK(false);
    }
}
```

this code outputs
```
    <Section name="Section1" filename="all_fail.cpp" line="10">
      <Section name="Section1.1" filename="all_fail.cpp" line="12">
        <Expression success="false" type="CHECK" filename="all_fail.cpp" line="13">
          <Original>
            false
          </Original>
          <Expanded>
            false
          </Expanded>
        </Expression>
        <OverallResults successes="0" failures="1" expectedFailures="0" skipped="false" durationInSeconds=<TIME>/>
      </Section>
      <OverallResults successes="0" failures="0" expectedFailures="0" skipped="false" durationInSeconds=<TIME>/>
    </Section>
```

Catch2:
```
    <Section name="Section1" filename="all_fail.cpp" line="10">
      <Section name="Section1.1" filename="all_fail.cpp" line="12">
        <Expression success="false" type="CHECK" filename="all_fail.cpp" line="13">
          <Original>
            false
          </Original>
          <Expanded>
            false
          </Expanded>
        </Expression>
        <OverallResults successes="0" failures="1" expectedFailures="0" skipped="false" durationInSeconds=<TIME>/>
      </Section>
      <OverallResults successes="1" failures="1" expectedFailures="0" skipped="false" durationInSeconds=<TIME>/>
    </Section>
```

Reply:
if we are counting the actual count for each section, then that needs to happen in the registry too? I don't see a way to do it in the `section_entry_checker` since it is only called on entry and exit.

Reply:
Yep; I've implemented this so the registry does the propagation of assert counts to all currently open sections.